### PR TITLE
Support deployment of Segmenter

### PR DESCRIPTION
**Pull request by RunningLeon**
Move from PR #559

## Motivation

Support Segmenter deployment with multiple backends

### models:
- [x] Segmenter Linear [configs/segmenter/segmenter_vit-s_linear_8x1_512x512_160k_ade20k.py](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/segmenter/segmenter_vit-s_linear_8x1_512x512_160k_ade20k.py), [ckpt](https://download.openmmlab.com/mmsegmentation/v0.5/segmenter/segmenter_vit-s_linear_8x1_512x512_160k_ade20k/segmenter_vit-s_linear_8x1_512x512_160k_ade20k_20220105_151713-39658c46.pth)
- [x] Segmenter Mask [configs/segmenter/segmenter_vit-s_mask_8x1_512x512_160k_ade20k.py](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/segmenter/segmenter_vit-s_mask_8x1_512x512_160k_ade20k.py), [ckpt](https://download.openmmlab.com/mmsegmentation/v0.5/segmenter/segmenter_vit-s_mask_8x1_512x512_160k_ade20k/segmenter_vit-s_mask_8x1_512x512_160k_ade20k_20220105_151706-511bb103.pth) 

### backends

- [x] onnxruntime
- [x] tensorrt
- [x] openvino
- [x] ncnn
- [x] torchscript

Note: pplnn: [Erf](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Erf) not supported.
## Modification

- [x] Add rewrites
- [x] Update docs in benchmark and support model list
- [x] Add unittest
- [x] Merge PR[#1642](https://github.com/open-mmlab/mmsegmentation/pull/1642) for deployment with shape not as 512x512.

## BC-breaking (Optional)
None

### CODE REVIEW

File: mmdeploy/pytorch/functions/chunk.py [hunk 1]
```diff
@@ -31,3 +32,19 @@ def chunk__ncnn(ctx, self, num_chunks: int, dim: int = 0) -> torch.Tensor:
     ]
 
     return output
+
+
+@FUNCTION_REWRITER.register_rewriter(
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/chunk.py:37]: Why do we need a chunk rewriter for torchscript?

**Comment by grimoire** [comment]
visualize failed with log:
```
2022-06-12:10:11:51 - root - ERROR - Error in execution: Non-zero status code returned while running Concat node. Name:'Concat_48' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/concat.cc:141 onnxruntime::common::Status onnxruntime::ConcatBase::PrepareForCompute(onnxruntime::OpKernelContext*, const std::vector<const onnxruntime::Tensor*>&, onnxruntime::Prepare&) const input_rank == reference_rank was false. Ranks of input data are different, cannot concatenate them. expected rank: 2 got: 3

```

File: mmdeploy/pytorch/functions/normalize.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    if p != 2:
+        return origin_func(input, p, dim=dim, eps=eps)
+    assert input.ndim in [3, 4]
+    if input.ndim == 3:
+        output = origin_func(
+            input.transpose(1, dim).unsqueeze(2), p=p, dim=1,
+            eps=eps).squeeze(2).transpose(1, dim)
+    else:
+        # input.ndim == 4:
+        assert input.shape[
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/normalize.py:34]: batch_size==1 could be assert for all branch.

File: mmdeploy/pytorch/functions/normalize.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    """Rewrite `normalize` for ncnn backend.
+
+    Make sure L2 norm on channel dim and be exported to ncnn correctly.
+    """
+    if dim < 0:
+        dim += input.ndim
+    assert dim != 0, 'Should not normalize on batch index'
+    origin_func = ctx.origin_func
+    if p != 2:
+        return origin_func(input, p, dim=dim, eps=eps)
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/normalize.py:26]: The broadcast problem should be exist even when p!=2?

File: mmdeploy/pytorch/functions/chunk.py (hunk 1 above)
Comment (RunningLeon) [reply mmdeploy/pytorch/functions/chunk.py:37]: @grimoire It breaks as follow, maybe there's better solution?
```
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmdeploy/mmdeploy/backend/torchscript/wrapper.py", line 49, in __init__
    self.ts_model = torch.jit.load(self.ts_model)
  File "/home/PJLAB/maningsheng/anaconda3/envs/pt1.8/lib/python3.7/site-packages/torch/jit/_serialization.py", line 161, in load
    cpp_module = torch._C.import_ir_module(cu, str(f), map_location, _extra_files)
RuntimeError: required keyword attribute 'chunks' is undefined
2022-06-14 10:58:09,677 - mmdeploy - ERROR - visualize torchscript model failed
```

**Comment by RunningLeon** [comment]
> visualize failed with log:
> 
> ```
> 2022-06-12:10:11:51 - root - ERROR - Error in execution: Non-zero status code returned while running Concat node. Name:'Concat_48' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/concat.cc:141 onnxruntime::common::Status onnxruntime::ConcatBase::PrepareForCompute(onnxruntime::OpKernelContext*, const std::vector<const onnxruntime::Tensor*>&, onnxruntime::Prepare&) const input_rank == reference_rank was false. Ranks of input data are different, cannot concatenate them. expected rank: 2 got: 3
> ```

@grimoire Which script you ran? could use regression_test tool to test segmenter.

File: mmdeploy/pytorch/functions/normalize.py (hunk 3 above)
Comment (RunningLeon) [reply mmdeploy/pytorch/functions/normalize.py:26]: Not sure. Since l2norm is common case and other form of norm is not met yet.

**Comment by grimoire** [comment]
> @grimoire Which script you ran? could use regression_test tool to test segmenter.

Works now with a static config.

**Review by tpoisonooo — COMMENTED** [review]
I have tried  ncnn ViT with RunningLeon:test_segmenter, LGTM.

**Review by grimoire — APPROVED** [review]

**Review by tpoisonooo — APPROVED** [review]
LGTM.

File: mmdeploy/pytorch/functions/linear.py [hunk 4]
```diff
@@ -25,7 +25,7 @@ def linear__ncnn(
 
     dim = input.dim()
 
-    if dim == 2 or dim == 3 and input.shape[0] == 1:
```
[the remark this answers is not in the document]
Comment (RunningLeon) [reply mmdeploy/pytorch/functions/linear.py:28]: did not remove it.

**Pull request merged**